<a href="https://colab.research.google.com/github/kanepeters/lis4693/blob/main/lab-2/Lab_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab 2: Text Pre-Processing using spaCy

In this lab assignment, we will learn to perform some basic text pre-processing using spaCy.

*Note: Before starting this lab assignment, please complete the Introduction to spaCy notebook*

## Leaning Objectives

In this exercise, you will:

- Load and process your own text file (transcript.txt)
- Split text into sentences
- Count words and sentences
- Find frequently used words
- Use spaCy’s PhraseMatcher to find specific phrases
- Understand the difference between blank pipelines and full pipelines

This exercise builds directly on concepts from discussed in the precursor notebook on "Introduction to spaCy".

## Install spaCy

In [2]:
!pip install spacy
!python -m spacy download en_core_web_sm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 71.5 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [3]:
import spacy

## Load and read your file from GitHub


In [4]:
import requests

url = "https://raw.githubusercontent.com/kanepeters/lis4693/refs/heads/main/lab-1/transcript.txt"
response = requests.get(url)
response.raise_for_status() # Raise an exception for HTTP errors
text = response.text

In [5]:
print("Number of characters:", len(text)) # print number of characters

print(text[:100])   # print first 100 characters

Number of characters: 3611
if you can correctly pronounce every
word in this poem you will be speaking
English better than 90 o


## Sentence Segmentation

Create a blank spaCy model and add sentencizer.

In [6]:
nlp = spacy.blank("en")
nlp.add_pipe("sentencizer")

doc = nlp(text)

sentences = list(doc.sents)

print("Number of sentences:", len(sentences))

print("\nFirst 3 sentences:")
for sent in sentences[:3]:
    print(sent)

Number of sentences: 2

First 3 sentences:
if you can correctly pronounce every
word in this poem you will be speaking
English better than 90 of the native
English speakers in the world after
trying the verses a Frenchman said he'd
prefer six months of hard labor to
reading six lines aloud
try them yourself
dearest creature in creation study
English pronunciation
I will teach you in my verse sounds like
corpse core hoarse and worse I will keep
you Susie busy make your head with heat
grow dizzy tear in eye your dress will
tear so shall I oh hear my prayer just
compare heart beard and herd dyes and
diet Lord and word sword and sword
retain and Britain mind the latter how
it's written now I surely will not play
you with such words as plaque and ague
but be careful how you speak say break
and stake but Bleak and streak cloven
oven how and low script receipt show
poem and toe hear me say devoid of
trickery daughter laughter and tip
sickery typhoid measles top sales Isles
Exile similes and r

## Word Count and Token Analysis

Let's count total words and unique words in your text file.

In [7]:
words = [token.text.lower() for token in doc if token.is_alpha]

print("Total words:", len(words))
print("Unique words:", len(set(words)))

Total words: 637
Unique words: 442


## Most Frequent Words

Let's find top 10 most frequent words in your file.

In [8]:
from collections import Counter

word_freq = Counter(words)

print("Top 10 most frequent words:")
for word, count in word_freq.most_common(10):
    print(word, count)

Top 10 most frequent words:
and 57
but 16
with 13
you 8
not 8
does 8
the 7
say 7
of 6
rhyme 6


## Using Full spaCy Pipeline

Now use the full model for better linguistic analysis.

In [9]:
nlp2 = spacy.load("en_core_web_sm")

doc2 = nlp2(text)

print("Named Entities:")
for ent in doc2.ents:
    print(ent.text, ent.label_)

Named Entities:
English LANGUAGE
90 CARDINAL
English NORP
Frenchman ORG
six months DATE
six CARDINAL
English NORP
Susie PERSON
Britain GPE
Bleak FAC
Mica PERSON
one CARDINAL
Laurel Gertrude German PERSON
Mallet Chalet ORG
Broad PERSON
Vape PERSON
River LOC
Grand PERSON
first ORDINAL
Ginger PERSON
Mirage ORG
Berry PERSON
Zephyr GPE
Senate ORG
Bull PERSON
George PERSON
Scenic Arabic Pacific Science ORG
Liberty Library PERSON
Mark PERSON
Constable ORG
Canal PERSON
Platt PERSON
Shay ORG
Arkansas GPE
Psalm Maria PERSON
Marine NORP
Italian NORP
Sally PERSON
Ava PERSON
Leisure PERSON
Heron Granary Canary ORG
Bass PERSON
Target ORG
seven CARDINAL
Ruffin PERSON
Stephen GPE
Turk PERSON
Cork PERSON
Stout PERSON
Groats GPE
Solace Gunnell PERSON
Islington GPE
Isle of Wight ORG


## PhraseMatcher

In [13]:
from spacy.matcher import PhraseMatcher

matcher = PhraseMatcher(nlp2.vocab, attr="LOWER")

phrases = ["English language", "English speakers", "rhyme"]

patterns = [nlp2(p) for p in phrases]

matcher.add("TECH_TERMS", patterns)

matches = matcher(doc2)

print("Matches found:")
for match_id, start, end in matches:
    print(doc2[start:end])
    print("Sentence:", doc2[start].sent)

Matches found:
English speakers
Sentence: if you can correctly pronounce every
word in this poem you will be speaking
English better than 90 of the native
English speakers in the world after
trying the verses a Frenchman said he'd
prefer six months of hard labor to
reading six lines aloud
try them yourself
dearest creature in creation study
English pronunciation
I will teach you in my verse sounds like
corpse core hoarse and worse I will keep
you Susie busy make your head with heat
grow dizzy tear in eye your dress will
tear so shall I oh hear my prayer just
compare heart beard and herd dyes and
diet Lord and word sword and sword
retain and Britain mind the latter how
it's written now I surely will not play
you with such words as plaque and ague
but be careful how you speak say break
and stake but Bleak and streak cloven
oven how and low script receipt show
poem and toe hear me say devoid of
trickery daughter laughter and tip
sickery typhoid measles top sales Isles
Exile similes and re

# EXCERCISE

Open a new Google Colab notebook and complete the tasks below. As you work, add brief explanations using the **Text (Markdown) cells** throughout your notebook to describe what you are doing.

1. Make a new folder named `lab-2` in your `lis4693` or `lis5693` repo on GitHub. **[1 Point]**

2. Complete the following tasks:


**TASK 1**: Load and read your `transcript.txt` file from lab-1 from GitHub repo directly **[1 Point]**

**TASK 2**: How many characters are in your transcript? Print the first 100 characters. **[1 Point]**

3611 characters in the transcript.

First 100 characters:
if you can correctly pronounce every
word in this poem you will be speaking
English better than 90 o

**TASK 3**: Perform sentence segmentation using the blank pipeline
  - How many sentences are in your transcript? **[0.5 Point]**
  
  Sentence Segmentation only produced 2 sentences.

  - Print the first 3 sentences **[0.5 Point]**

(the entire transcript was parsed as large sentences with sparse punctuation)
  
if you can correctly pronounce every
word in this poem you will be speaking
English better than 90 of the native
English speakers in the world after
trying the verses a Frenchman said he'd
prefer six months of hard labor to
reading six lines aloud
try them yourself
dearest creature in creation study
English pronunciation
I will teach you in my verse sounds like
corpse core hoarse and worse I will keep
you Susie busy make your head with heat
grow dizzy tear in eye your dress will
tear so shall I oh hear my prayer just
compare heart beard and herd dyes and
diet Lord and word sword and sword
retain and Britain mind the latter how
it's written now I surely will not play
you with such words as plaque and ague
but be careful how you speak say break
and stake but Bleak and streak cloven
oven how and low script receipt show
poem and toe hear me say devoid of
trickery daughter laughter and tip
sickery typhoid measles top sales Isles
Exile similes and revials Scholar vicar
and cigar solar Mica war and far
one anemone kitchen like and laundry
Laurel Gertrude German wind and mine
seen melcomony mankind
Billet does not rhyme with ballet
bouquet wallet Mallet Chalet blood and
flood are not like food nor is mold like
should and would viscous VI count load
and Broad toward to forward to reward
and your pronunciations okay when you
correctly say croquet rounded wounded
grief and sieve friend and fiend alive
and live I Vape previc famous clamor and
enamor rhyme with hammer River rival
tomb bomb comb dull and roll and some
and home
stranger does not rhyme with anger
neither does devour with clanger
Souls but foul haunt but aren't font
front won't want Grand and Grant
shoes goes does now first say finger and
then singer Ginger linger
real Zeal move goes gouge and Gage
marriage foliage Mirage and age
query does not rhyme with very nor does
Fury sound like Berry
dust lost post and doth cloth loath job
knob bosom transom oath though the
differences seem little we say actual
but vital
refer does not rhyme with defer fever
does and Zephyr heifer mint pint Senate
and sedate dull Bull and George ate late
Scenic Arabic Pacific Science conscience
scientific
Liberty Library heave and Heaven Rachel
ake mustache 11.
we say hallowed but
aloud people leopard toad but vowed Mark
the differences moreover between mover
cover clover
leeches breaches wise precise chalice
but police and lies
camel Constable unstable principal
disciple label petal panel and Canal
wait surprise Platt promise pal whim and
storm Shay's chaos chair Senator
spectator mayor tall but our and sucker
for gas Alas and Arkansas
C idea career area Psalm Maria but
malaria youth south southern cleanse and
clean Doctrine at turpentine Marine
compare alien with Italian dandelion and
Battalion
Sally with Ally yay Yi
[Music]
way and key
say Ava but ever fever neither Leisure
skein deceiver Heron Granary Canary
crevice and a device and Airy face but
preface not a face phlegm phlegmatic ass
glass Bass large but Target gin give
virgin
ought out joust and scour scourging hear
but earn and wear and tear do not rhyme
with hear but air
seven is right but so is even hyphen
Ruffin nephew Stephen
monkey donkey Turk and jerk ask grasp
wasp and Cork and work
pronunciation think of psyche is a
paling Stout and spiky wanted to make
you lose your wits writing Groats and
saying grits it's a dark Abyssal tunnel
strewn with stones stowed Solace Gunnell
Islington and Isle of Wight housewife
verdict and indict
finally which rhymes with enough
though through plow or dough or cough
hiccup has the sound of cup
my advice is to give up
[Music]

**TASK 4**: Perform Word Count and Token Analysis
  - What is the total number of words? **[0.5 Point]**

Total words: 637

  - What is the number of unique words? **[0.5 Point]**

Unique words: 442

**TASK 5**: Find Most Frequent Words
  - What is the most frequent word? **[0.5 Point]**
  
The most frequent word is "and" (57 instances).

  - Why do you think this word appears frequently? **[1 Point]**

The video recites a poem that combines various homophones, homographs, and other similar-sounding or similarly-spelled words to demonstrate the complexity of the English language. In doing so, the word "and" is commonly used to connect such words.

**TASK 6**: Run full spaCy pipeline
  - How many named entities were found? **[0.5 Point]**

53 named entities were found.

  - What types of entities appear? (PERSON, ORG, DATE, etc.) **[0.5 Point]**

Types of entities include: PERSON, ORDINAL, ORG, GPE, NORP

**TASK 7**: Use PhraseMatcher **[1 Point]**

When you selected your YouTube video in Lab-1, what topic or subject were you interested in? Based on that topic, identify three specific phrases that are directly relevant to your search. For example, if your video was about information retrieval, relevant phrases might include "information retrieval," "text mining," and "data science."

Make sure the video you selected in Lab-1 was clearly related to your chosen topic and was the same video for which you downloaded the transcript in Lab-1.

I selected this YouTube video without any particular regard for a topic or subject, but for the purpose of this response it could be said that this video is most closely related with topics such as language and particularly the English language. As the video content features a fairly unique poem, the presence of specific phrases might be limited to ones such as "English language" or "English speakers."

**TASK 8**: At the end of your Colab notebook, create a new text cell and write a brief reflection for this assignment in a few sentences addressing the following **[2 Points]**:
 - What went well?

The pipeline effectively counted up several useful figures, such as the total word count, the number of unique words, and the frequency of certain words in a manner that would otherwise be difficult to tally manually.

 - What did not go well or what challenges you encountered?

The transcript produced in Lab 1, which was sourced from the YouTube video I selected, seems to be mainly derived from a video that had its captions generated automatically by YouTube. In particular, it exhibits certain qualities that are common among such similarly generated captions; namely, the transcript itself is utterly lacking in punctuation that would otherwise be present for videos that had their captions produced manually. Also, the auto-generated captions aren't necessarily accurate; this can be visually confirmed by viewing the video itself which displays the actual poem as it is narrated over the course of the video. In short, some words are transcribed incorrectly. This phenomenon itself is to be expected since the topic of the video itself deals with words such as homophones or homographs, as well as words with pronunciation that are likely to be misrepresented by the auto-captions.

Due to these factors, some of the results produced by spaCy are clearly questionable. Notably, the package considers the transcript itself to be made up of a few extremely long sentences due to the scarcity of punctuation marks. Also, as many words in the transcript are capitalized, spaCy liberally interprets many words as named entities in a spurious manner. Additionally, some named entities get combined into a single named entity due to the auto-captioning's omission of commas.

3. Push your Google Colab file to your `lab-2` GitHub repo from Colab. *No points will be given if you upload it to GitHub directly!* **[1 Point]**
4. Share the link to your `lab-2` GitHub repo for this lab assignment on CANVAS for credit. **[0.5 Point]**
